Segmentation fault during tests on FreeBSD and Mac OS X

Tested on FreeBSD 13/amd64 and macOS/arm64, both fail at the same point

Here's the log from FreeBSD:

3 warnings generated.
cc -g -W -Wall dumper.o pdp10-opc.o info.o dis.o symbols.o timing.o timing_ka10.o timing_ki10.o memory.o weenix.o libwords.a -o dumper
sh check.sh && touch check
OK: ts.obs.dasm
OK: ts.ksfedr.dasm
Segmentation fault (core dumped)
gmake: *** [Makefile:125: check] Error 139

Running under a debugger:
375   772000   400000,,210175  a---r 175
376   774000   400000,,210176  a---r 176
377   776000   400000,,210177  a---r 177
Process 7953 stopped
* thread #1, name = 'dis10', stop reason = signal SIGSEGV: invalid address (fault address: 0xffffffffffffeff8)
    frame #0: 0x00000008003c162f libc.so.7`memcpy + 607
libc.so.7`memcpy:
->  0x8003c162f <+607>: rep    movsq	(%rsi), %es:(%rdi)
    0x8003c1632 <+610>: cld
    0x8003c1633 <+611>: movq   %rdx, %rcx
    0x8003c1636 <+614>: andb   $0x7, %cl

and:
(lldb) bt all
* thread #1, name = 'dis10', stop reason = signal SIGSEGV: invalid address (fault address: 0xffffffffffffeff8)
  * frame #0: 0x00000008003c162f libc.so.7`memcpy + 607
    frame #1: 0x0000000000219635 dis10`purify_memory + 325
    frame #2: 0x000000000021aad3 dis10`read_pdump + 1123
    frame #3: 0x0000000000210733 dis10`main + 595
    frame #4: 0x00000000002102f0 dis10`_start + 256

--- a/memory.c
+++ b/memory.c
@@ -55,7 +55,7 @@ init_memory (struct pdp10_memory *memory)
   memory->current_address = 0;
 }
 
-static void
+static struct pdp10_area *
 insert_area (struct pdp10_memory *memory, int i)
 {
   memory->areas++;
@@ -68,6 +68,7 @@ insert_area (struct pdp10_memory *memory, int i)
 
   memmove (&memory->area[i+1], &memory->area[i],
 	   (memory->areas - i - 1) * sizeof (struct pdp10_area));
+  return &memory->area[i];
 }
 
 int
@@ -103,9 +104,7 @@ add_memory (struct pdp10_memory *memory, int address, int length, word_t *data)
       return 0;
     }
 
-  insert_area (memory, i);
-
-  area = &memory->area[i];
+  area = insert_area (memory, i);
   area->start = address;
   area->end = address + length;
   area->flags = 0;
@@ -179,7 +178,7 @@ purify_memory (struct pdp10_memory *memory, int address, int length)
       if (area->start < i)
 	{
 	  /* Impure area needs to split off first part. */
-	  insert_area (memory, area - memory->area);
+	  area = insert_area (memory, area - memory->area);
 	  area->end = i;
 	  data = area->data + area->end - area->start;
 
@@ -194,7 +193,7 @@ purify_memory (struct pdp10_memory *memory, int address, int length)
       if (area->end > end)
 	{
 	  /* Impure area needs to split off last part. */
-	  insert_area (memory, area - memory->area);
+	  area = insert_area (memory, area - memory->area);
 	  area->end = i;
 	  area->flags |= MEMORY_PURE;
 

